### PR TITLE
fix(web): fix keyboard-processing bugs uncovered by stricter TS settings 🔩

### PR DESCRIPTION
**Pull request by jahorton**
Originally part of #11424, these changes address bugs that were uncovered thanks to stricter TS settings.  There are three in total, but they're in the same region of code and are all quite small, so I lumped 'em together.

@keymanapp-test-bot skip

### CODE REVIEW

File: common/web/keyboard-processor/src/keyboards/defaultLayouts.ts [hunk 1]
```diff
@@ -61,7 +61,10 @@ export class Layouts {
   static readonly dfltText='`1234567890-=\xA7~~qwertyuiop[]\\~~~asdfghjkl;\'~~~~~?zxcvbnm,./~~~~~ '
     +'~!@#$%^&*()_+\xA7~~QWERTYUIOP{}\\~~~ASDFGHJKL:"~~~~~?ZXCVBNM<>?~~~~~ ';
 
-  static readonly DEFAULT_RAW_SPEC = {'F':'Tahoma', 'BK': Layouts.dfltText} as const;
+  // The function baked into keyboards by the current Web compiler creates an
+  // array of single-char strings for BK. Refer to
+  // developer/src/kmc-kmn/src/kmw-compiler/visual-keyboard-compiler.ts.
+  static readonly DEFAULT_RAW_SPEC = {'F':'Tahoma', 'BK': Layouts.dfltText.split('')} as const;
```
Comment (jahorton) [inline common/web/keyboard-processor/src/keyboards/defaultLayouts.ts:67]: From an _old_ keyboard (viewable [here](https://github.com/keymanapp/keyman/blob/c8d8e4187aa20f7db7f22c8fa7ccd21a04a2615b/web/src/test/manual/web/us-1.0.js)):

```javascript
BK: new Array("`","1","2","3","4","5","6","7","8","9","0","-","=","","","","q","w","e","r","t","y","u","i","o","p","[","]","\\","","","","a","s","d","f","g","h","j","k","l",";","'","","","","","","\\","z","x","c","v","b","n","m",",",".","/","","","","",""," ","~","!","@","#","$","%","^","&","*","(",")","_","+","","","","Q","W","E","R","T","Y","U","I","O","P","{","}","|","","","","A","S","D","F","G","H","J","K","L",":","\"","","","","","","|","Z","X","C","V","B","N","M","<",">","?","","","","",""," ","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","",""," ")
```

From the output generated by modern compilers when run live (via `khmer_angkor`):

> (130) ['«', '១', '២', '៣', '៤', '៥', '៦', '៧', '៨', '៩', '០', 'ឥ', 'ឲ', '', '', '', 'ឆ', '', '', 'រ', 'ត', 'យ', '', '', '', 'ផ', '', 'ឪ', 'ឮ', '', '', '', '', 'ស', 'ដ', 'ថ', 'ង', 'ហ', '', 'ក', 'ល', '', '', '', '', '', '', '', '', 'ឋ', 'ខ', 'ច', 'វ', 'ប', 'ន', 'ម', '', '។', '', '', '', '', '', '', '​', '»', '!', 'ៗ', '"', '៛', '%', '', '', '', '(', ')', '', '=', '', '', '', 'ឈ', '', '', 'ឬ', 'ទ', '', '', '', '', 'ភ', '', 'ឧ', 'ឭ', '', '', '', '', '', 'ឌ', …]

It _is_ technically OK in JavaScript to index a string as if it were an array, which is why things have been working.  TS strictness settings don't like it, though.

File: common/web/keyboard-processor/src/text/codes.ts [hunk 2]
```diff
@@ -86,7 +86,7 @@ const Codes = {
     [')!@#$%^&*(',':+<_>?~', '{|}"']
   ],
 
-  isKnownOSKModifierKey(keyID: string): boolean {
+  isKeyNotCorrected(keyID: string): boolean {
```
Comment (jahorton) [inline common/web/keyboard-processor/src/text/codes.ts:89]: I changed the name for two reasons:

1. This highlights the limited uses of the method.
2. Given that the only use is to determine what keys we don't fat-finger correct, and that K_TAB + K_TABFWD are being included... well, K_TAB isn't a modifier key.

File: common/web/keyboard-processor/src/text/codes.ts [hunk 3]
```diff
@@ -98,13 +98,6 @@ const Codes = {
         if(Codes.keyCodes[keyID] >= 50000) { // A few are used by `sil_euro_latin`.
           return true; // is a 'K_' key defined for layer shifting or 'control' use.
         }
-        // Refer to text/codes.ts - these are Keyman-custom "keycodes" used for
-        // layer shifting keys.  To be safe, we currently let K_TABBACK and
-        // K_TABFWD through, though we might be able to drop them too.
-        const code = Codes[keyID];
-        if(code > 50000 && code < 50011) {
-          return true;
-        }
```
Comment (jahorton) [inline common/web/keyboard-processor/src/text/codes.ts:107]: Once I stopped and looked at this... "wait, why do we have two near-identical checks"?  And... "when did this happen?"

The answer to the latter question: it happened in #7242.  I don't know what I was thinking then.  It's clear we should only have one of the two.

Granted, settling on the one that I did kind of implies we should also ignore `K_TAB` for consistency.  (I haven't made this change yet, though.)

File: common/web/keyboard-processor/src/text/kbdInterface.ts [hunk 4]
```diff
@@ -938,6 +938,7 @@ export default class KeyboardInterface extends KeyboardHarness {
     if(systemId == SystemStoreIDs.TSS_LAYER && this.activeDevice.touchable) {
       // Denote the changed store as part of the matched rule's behavior.
       this.ruleBehavior.setStore[systemId] = strValue;
+      return true;
```
Comment (jahorton) [inline common/web/keyboard-processor/src/text/kbdInterface.ts:941]: Note line 929, which says we should have been doing this all along.

**Review by mcdurdin — APPROVED** [review]
I think this should be three separate pull requests -- they are three unrelated bugs.

The three bug fix titles could have been:
1. fix(web): `[keyboard].KV.BK` should always be an array but default fallback was providing a string
2. refactor(web): Remove dead code in `isKnownOSKModifierKey` and rename it to `isKeyNotCorrected`
3. fix(web): always return true or false for `KSETS` (`setStore`)

In my analysis, none of the three changes have any end-user impact.
1. as you noted, `[keyboard].KV.BK` as string or array had basically the same outcome, but would be a trap for future maintainers
2. there is a question you ask regarding `isKeyNotCorrected` which probably could turn into its own discussion:
    > Granted, settling on the one that I did kind of implies we should also ignore K_TAB for consistency. (I haven't made this change yet, though.)
3. keyboards do not use the return value from `KSETS`: 
  https://github.com/keymanapp/keyman/blob/ab4fcef9f7a8af90d3eee27d9e06ba688422b86c/developer/src/kmc-kmn/src/kmw-compiler/javascript-strings.ts#L883-L884

Generally, including `Not` in a function name is a [code](https://stackoverflow.com/a/52689404/1836776) [smell](https://softwareengineering.stackexchange.com/a/196837). A better name may be `isFrameKey`?

Apart from that, LGTM. I guess I will approve this PR this time, but let's try and get into the habit of fixing one bug at a time in a PR.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.47-alpha](https://keyman.com/downloads/releases/alpha/18.0.47)

**Pull request merged**